Asynchronous download button for ORA2 data

dylanrhodes opened the pull request:
This pull request implements a management command and instructor dashboard button which gather all open assessment response data for a given course in a user friendly format and output it as a .csv.  It does so by submitting a task to the celery queue and waiting for it to upload the result to S3 in an asynchronous fashion.  Data collection is performed via a templatized, raw SQL query (admittedly a bit unorthodox), the original version of which may be found <a href="https://gist.github.com/wedaly/f17ba1d2fffae65545c8">here</a>.  Also included are unit tests for the functions added in various locations.  Feedback is welcome and encouraged @jbau @sarina as well as tagging more people who should have a look.  Thanks!

sarina commented:
@dylanrhodes happy 2015 :confetti_ball: !!

Let me know if you need any help responding to comments, etc on this PR. This is a feature we're greatly excited about!

sarina commented:
No worries! Thanks for letting me use your desk!

On Mon, Jan 5, 2015 at 3:37 PM, Dylan Rhodes <notifications@github.com>
wrote:

> Hey @sarina <https://github.com/sarina> I'm back in the office as of
> today, so I'll respond to all of the open comments and let you know about
> further steps ASAP. Thanks for bearing with me over the holiday :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/5987#issuecomment-68776279>.
>

sarina commented:
@dylanrhodes - how's this going? Is this ready to be code reviewed?

dylanrhodes commented:
@sarina I think that would be a good next move.  There haven't been many logical changes since the last round of reviews, mainly switching the HTTP response code for asynchronous report generation requests to 202.

sarina commented on lms/djangoapps/instructor/management/commands/collect_ora2_data.py:
```diff
@@ -0,0 +1,64 @@
+"""
+Command to retrieve all Ora2 data for a course in a .csv.
```
Nit: "ORA" is an acronym so please capitalize its use: "ORA2"

sarina commented on lms/djangoapps/instructor/utils.py:
```diff
@@ -36,3 +39,120 @@ def get_module_for_student(student, usage_key, request=None):
+    """
+    cursor = get_read_replica_cursor_if_available(db)
+
+    #Syntax unsupported by other vendors such as SQLite test db
```
Nit: please provide a space between `#` and the start of a comment

sarina commented on lms/djangoapps/instructor/utils.py:
```diff
@@ -36,3 +39,120 @@ def get_module_for_student(student, usage_key, request=None):
+    """
+
+    # pylint: disable=invalid-name
+    RAW_QUERY = """
```
I wonder if this should be defined in its own text file, to make future editing of it more straightforward. Then you could have this function read in the file and shove the file contents into the `RAW_QUERY` variable.

dylanrhodes replied:
I thought about doing it that way, but reading in a text file and directly executing its contents against the database seems unsafe to me.  I think that there are fewer points of failure and less potential for inadvertent SQL injection this way.  I'm open to a discussion about it, though.

sarina replied:
That sounds reasonable. I'm going to ask @stephensanchez to weigh in here. Steve, could you just look at the one question posted above? Usman and his team are going to look at this PR in their sprint this week, but this seems like a bit more of an architectural question.

stephensanchez replied:
@sarina thanks for the tag.  I have some larger questions overall, and I'll comment at the bottom of this PR, but if we were to go ahead with this implementation I'd agree that in-lining the raw SQL makes sense, but I'd recommend making a common abstract class where this raw query is just one single implementation of how to get ORA2 data, and other implementations could use Django Models, or an entirely different query.

This still may run the risk of an injection, but only at a configuration level.

sarina commented on lms/djangoapps/instructor/views/api.py:
```diff
@@ -1644,12 +1644,10 @@ def calculate_grades_csv(request, course_id):
         instructor_task.api.submit_calculate_grades_csv(request, course_key)
         success_status = _("Your grade report is being generated! You can view the status of the generation task in the 'Pending Instructor Tasks' section.")
-        return JsonResponse({"status": success_status})
+        return JsonResponse({"status": success_status}, status=202)
```
why does this return 2 statuses here?

dylanrhodes replied:
It doesn't actually return two statuses; there's just an unfortunate naming collision in the JsonResponse class.  The status field within the dictionary corresponds to the status message which is displayed to the user upon acceptance of their request for report generation.  The status parameter is the HTTP response code which is returned to their browser.

sarina replied:
Cool, thanks for the clarification. I'll admit I don't know ORA super well; my job is to run through prs and flag general issues and questions.

sarina commented on lms/djangoapps/instructor/views/api.py:
```diff
@@ -1717,6 +1715,30 @@ def extract_user_info(user):
+    course_key = SlashSeparatedCourseKey.from_deprecated_string(course_id)
+    try:
+        instructor_task.api.submit_ora2_request_task(request, course_key)
+        success_status = _("The ORA2 responses report is being generated.")
```
"Responses report" sounds odd to me. Maybe "response report"?

sarina commented:
Codewise this looks pretty good. I don't know ORA very well but made some general nits for style and clarity. I'm going to hand this off to the appropriate engineering team for review.

sarina commented:
Heads up @mhoeber - this has documentation impact. There's a small number of user facing strings that should be reviewed.

mhoeber commented:
@sarina  thanks, have this on board:  https://openedx.atlassian.net/browse/DOC-1515

sarina commented:
@dylanrhodes : FYI this pull request is scheduled to be reviewed in the next two weeks. Please plan to be available to respond to review comments.

Right now, your PR is still mergeable, so it does not need a rebase. However, could I please urge you to take a second look at your [diff-coverage report](https://jenkins.testeng.edx.org/job/edx-platform-report/8383/Diff_Coverage_Report/)? It reports 79% coverage. We strive for 95+% coverage as much as possible.

sarina commented:
@dylanrhodes also, there's a few comments that are as-of-yet not responded to. Could you please take a look at this, too? cc @jbau

explorerleslie commented:
@dylanrhodes @sarina FYI I am prioritizing this PR review for the team for the sprint starting this Wednesday. @dylanrhodes please try to have all outstanding issues mentioned in @sarina 's review resolved.

dylanrhodes commented:
@sarina I am available to look at this PR this week.  I've just pushed a commit which responds to all of your open comments save the diff-coverage report.  The PR is admittedly missing test cases for the collect_ora2_data management command (although it just writes a .csv, nothing special).  Those tests are forthcoming.  

@explorerleslie I believe that the PR as a whole is ready for an in-depth review.  We've been running a nearly identical version of the ORA2 report button here at Stanford since November without issues, so hopefully it will be a smooth process.

stephensanchez commented on common/djangoapps/util/query.py:
```diff
@@ -7,3 +8,15 @@ def use_read_replica_if_available(queryset):
+    Returns cursor to read_replica or default if not available
+    """
+    try:
+        cursor = db.connections['read_replica'].cursor()
```
ah, you are looking for the read_replica, great! That was my main concern, and whether or not this conflicted with getting ORA2 data into the data package from Analytics @shnayder 

I think for production, we'd want to force the use of the read replica, and never let this use the live DB.

shnayder replied:
Nope, no conflict -- even when we get ORA2 data into the data packages, we'll still need a way for course teams to get it directly, and in a less raw format than the underlying tables.

sarina commented:
@dylanrhodes not sure if I've sent this to you yet: 
Hi! You may notice we're using a new Jenkins server. So, how do you read your build output?

Click on the build mark (red x or yellow circle) next to your commit, to be brought to the build page. From there, click on the "Build Graph" link on the left hand side. You'll see a flow chart diagram. Click on the title of the second box ("edx-platform-all-tests-pr #(build-number): (commit-hash)"), which represents the actual test build, to be brought to the test page.

On the test page, you can see what shard(s) passed or failed. Click on the failing shards, represented by red circles, to see what failed in your build. 

If you see a green check, your build passed! Clicking on it will bring you to a report page where you can view the output of diff-coverage and diff-quality reports. As a reminder, we expect coverage to be 95% or higher as much as possible, and diff-quality to be 100% (you can run this locally using the `paver run_quality` command).

Looks like your quality shard is failing. Make sure diff-quality is 100%.

dylanrhodes commented:
Hey @sarina my local linter must have missed something - I must not have checked in the new test file before I ran it or something.  I have the new Jenkins display figured out, though, so I'll keep at it if it doesn't pass with the whitespace fixes I just pushed.

sarina commented:
@dylanrhodes cool, looks good! Someone from the ORA team will review this in the next sprint cycle (started yesterday, and goes for two weeks).

sarina commented on lms/djangoapps/instructor/tests/test_api.py:
```diff
@@ -250,13 +251,18 @@ def test_staff_level(self):
+            expected_status = 200
+
+            if endpoint in ['calculate_grades_csv', 'get_ora2_responses']:
+                expected_status = 202
```
This line is reported as uncovered. It's a bit concerning when test lines don't get hit. Can you double-check this? https://build.testeng.edx.org/job/edx-platform-report-pr/380/Diff_Coverage_Report/

symbolist commented on lms/djangoapps/instructor_task/tasks.py:
```diff
@@ -166,3 +167,13 @@ def cohort_students(entry_id, xmodule_instance_args):
+
+
+@task(base=BaseInstructorTask, routing_key=settings.ORA2_RESPONSES_DOWNLOAD_ROUTING_KEY)  # pylint: disable=not-callable
+def get_ora2_responses(entry_id, xmodule_instance_args):
```
Since this exports more than just responses, I think a better name would be `export_ora2_data`.

symbolist commented on lms/djangoapps/instructor_task/api.py:
```diff
@@ -250,3 +253,15 @@ def submit_cohort_students(request, course_key, file_name):
+    """
+    AlreadyRunningError is raised if an ora2 report is already being generated.
+    """
+    task_type = 'ora2_responses'
```
And for consistency we can name the function `submit_export_ora2_data` and the `task_type` `export_ora2_data`.

symbolist commented on lms/djangoapps/instructor_task/tasks_helper.py:
```diff
@@ -738,3 +740,64 @@ def cohort_students_and_upload(_xmodule_instance_args, _entry_id, course_id, tas
+        return UPDATE_STATUS_FAILED
+
+    timestamp_str = start_time.strftime('%Y-%m-%d-%H%M')
+    course_id_string = urllib.quote(course_id.to_deprecated_string().replace('/', '_'))
```
Instead of `course_id.to_deprecated_string()`, you can use `unicode(course_id)`.

symbolist commented on lms/djangoapps/instructor_task/tasks_helper.py:
```diff
@@ -738,3 +740,64 @@ def cohort_students_and_upload(_xmodule_instance_args, _entry_id, course_id, tas
+    num_total = 1
+    curr_step = "Collecting responses"
+
+    def update_task_progress():
```
Can you explain why we need to do this instead of just using the `TaskProgress` object like in the other tasks?

symbolist commented on lms/templates/instructor/instructor_dashboard_2/data_download.html:
```diff
@@ -39,8 +39,10 @@
-    <p><input type="button" name="calculate-grades-csv" value="${_("Generate Grade Report")}" data-endpoint="${ section_data['calculate_grades_csv_url'] }"/></p>
+    <p>
+      <input type="button" name="calculate-grades-csv" class="async-report-btn" value="${_("Generate Grade Report")}" data-endpoint="${ section_data['calculate_grades_csv_url'] }"/>
+      <input type="button" name="ora2-response-btn" class="async-report-btn" value="${_("Download ORA2 Responses")}" data-endpoint="${ section_data['get_ora2_responses_url'] }"/>
```
To be consistent with the names of other input fields, this should be `export-ora2-data`.

sarina replied:
also to be consistent, the text should be "Generate ORA2 Responses", not "Download".

Download implies that clicking a button immediately begins a download, not kick off a background task. So, please change this.

sarina replied:
Also, are you showing this for all courses, even those that don't use ORA2? I wonder if it's possible to flag this so that it only shows for courses that make use of ORA2.

symbolist replied:
For this we will have to fetch all the items from mongo and check if any is of the right type. @explorerleslie Do we need to do this?

sarina replied:
Well if it's not feasible that's fine. I wasn't sure if there was an easy way to tell or not. Thanks for the info

symbolist commented on lms/static/coffee/src/instructor_dashboard/data_download.coffee:
```diff
@@ -107,17 +108,20 @@ class DataDownload
+          if e.target.name == 'calculate-grades-csv'
+            @$grades_request_response_error.text gettext("Error generating grades. Please try again.")
+          else if e.target.name == 'ora2-response-btn'
+            @$grades_request_response_error.text gettext("Error getting ORA2 responses. Please try again.")
```
`"Error getting ORA2 responses. Please try again."`

sarina replied:
@symbolist that's what is written in the code.

Should it say maybe "generating" as opposed to "getting"? If this is a background task run by Celery, I would say this should say 'generating' not 'getting' so as to be consistent with the other existing task.

symbolist replied:
Yes. Sorry, I copied the string but forgot to edit in my suggestion. I think all user facing strings should get a review from product so I haven't commented on any others.

symbolist replied:
Yes. Sorry, I copied the string but forgot to edit in my suggestion. I think all user facing strings should get a review from product so I haven't commented on any others.

symbolist replied:
Yes. Sorry, I copied the string but forgot to edit in my suggestion. I think all user facing strings should get a review from product so I haven't commented on any others.

symbolist commented on lms/envs/common.py:
```diff
@@ -1746,6 +1746,15 @@
+
+ORA2_RESPONSES_DOWNLOAD = {
+    'STORAGE_TYPE': 'localfs',
+    'BUCKET': 'edx-grades',
```
You are using a different `ROOT_PATH`; why not also a different `BUCKET`?

symbolist commented on lms/djangoapps/instructor/management/commands/collect_ora2_data.py:
```diff
@@ -0,0 +1,67 @@
+            except InvalidKeyError:
+                raise CommandError("The course ID given was invalid")
+
+        file_name = ("%s-ora2.csv" % course_id).replace("/", "-")
```
Can we have the timestamp in the filename too? It makes running the command multiple times simpler.

symbolist commented:
@dylanrhodes The code looks pretty good! I only have some minor suggestions.

There is one thing that I think needs a bit of cleanup though. Currently, a number of different expressions are being used for this functionality in different places e.g. "aggregate ora2 responses", "ORA2 response report", "get_ora2_responses", "if an ora2 report is already being generated", "ORA2_RESPONSES_DOWNLOAD".

It would be great if we referred to this functionality in a consistent way everywhere. For example "export ora2 data" and its variants would work nicely.

@explorerleslie @stephensanchez @ormsbee I do have a question about the data format. ORA2 data has 3 levels of nesting and exporting it into a csv gets us something like this: https://gist.github.com/dylanrhodes/d404168f3599c79c2dcb. This data is generated using this [complex concatenation](https://github.com/Stanford-Online/edx-platform/blob/dylanrhodes/upstream_ora2_button/lms/djangoapps/instructor/utils.py#L78-L109) in sql queries and it is not clear to me how parseable the result is/will remain in the future.

Can we consider other options?

symbolist commented:
@dylanrhodes I want to add one more thing: I think it makes sense to move all the ORA specific code and the django command to the edx-ora2 repo and just have the instructor task logic here. This way if any ORA models are edited, the queries can be edited as well together.

explorerleslie commented:
@symbolist good comment on the format of the spreadsheet. Agreed that this isn't the ideal format, but it is the same format we're manually providing to clients via a devops management command right now. So, I'd like to get this information into the instructor dashboard in some form first, and then tweak the format later.

symbolist commented:
I created a sandbox from the branch but am getting the following error when trying to run the task.
```
Feb  9 15:39:03 ip-192-168-0-72 [service_variant=lms][celery.worker.consumer][env:sandbox] ERROR [orareport  8613] [consumer.py:574] - Received unregistered task of type u'instructor_task.tasks.get_ora2_responses'.
The message has been ignored and discarded.

Did you remember to import the module containing this task?
Or maybe you are using relative imports?
More: http://docs.celeryq.org/en/latest/userguide/tasks.html#names

The full contents of the message body was:
{u'utc': True, u'chord': None, u'args': [2, {u'request_info': {u'username': u'staff', u'ip': u'127.0.0.1', u'host': u'orareport.m.sandbox.edx.org', u'agent': u'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_9_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/40.0.2214.94 Safari/537.36'}, u'xqueue_callback_url_prefix': u'http://orareport.m.sandbox.edx.org', u'task_id': u'4c07e29a-059e-4002-99f8-c8c8a2e6e10d'}], u'retries': 0, u'expires': None, u'task': u'instructor_task.tasks.get_ora2_responses', u'callbacks': None, u'errbacks': None, u'taskset': None, u'kwargs': {}, u'eta': None, u'id': u'4c07e29a-059e-4002-99f8-c8c8a2e6e10d'} (601b)
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/worker/consumer.py", line 438, in on_task_received
    strategies[name](message, body, message.ack_log_error)
KeyError: u'instructor_task.tasks.get_ora2_responses'
```

dylanrhodes commented:
Hey everyone, thanks for looking at the PR.  I've been asked to prioritize a few different projects this sprint in light of the staff changes here, so I probably won't get to this for a little while.

jinpa commented:
It sounds like people are starting to request some substantial changes here.  @dylanrhodes  unfortunately won't be able to work on this anymore, because he has other pressing projects.  He's made a number of changes in response to comments since the original submission in November, but isn't planning on working more on this at this point.  Would you like him to close this PR, or would you like to merge it?  We've been running this successfully here for several months, and it meets a customer need.

symbolist commented:
@jinpa @dylanrhodes Thanks for letting us know! The TNL team will take care of making the requested changes. You can follow the progress on this new PR: https://github.com/edx/edx-platform/pull/6975
